### PR TITLE
Add rclc_parameter Quality Declaration

### PR DESCRIPTION
**Pull request by pablogs9**
Closes https://github.com/ros2/rclc/issues/137

### CODE REVIEW

**Comment by pablogs9** [comment]
@mergify backport galactic

File: rclc_parameter/QUALITY_DECLARATION.md [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+Most features in `rclc_parameter` have corresponding tests which simulate typical usage, and they are located in the [`test`](./test) directory.
+New features are required to have tests before being added.
+
+### Public API Testing [4.ii]
+
+Each part of the public API has tests, and new additions or changes to the public API require tests before being added. The tests aim to cover both typical usage and corner cases, but are quantified by contributing to code coverage.
+
+### Coverage [4.iii]
+
+Coverage checks for `rclc_parameter` have not been implemented.
```
Comment (JanStaschulat) [inline rclc_parameter/QUALITY_DECLARATION.md:105]: @pablogs9 codecov is enabled for all packages of rclc: That includes also rclc_parameters. 
e.g. https://codecov.io/gh/ros2/rclc/tree/galactic/ros_ws/src/0oshwe2aqpon/rclc/rclc_parameter/src/rclc_parameter

Comment (JanStaschulat) [inline rclc_parameter/QUALITY_DECLARATION.md:105]:
```suggestion
The coverage report is available in the [../README.md](../README.md).

`rclc_parameter` checks the coverage of every commit. Last coverage assessment can be seen in [Codecov](https://app.codecov.io/gh/ros2/rclc/commits).
```

**Review by JanStaschulat — CHANGES_REQUESTED** [review]
@pablogs9 @norro We need to test rclc_parameter on Windows first before we can add this quality declaration for this package.

File: rclc_parameter/QUALITY_DECLARATION.md (hunk 1 above)
Comment (pablogs9) [reply rclc_parameter/QUALITY_DECLARATION.md]: Thanks

**Comment by norro** [comment]
With the branch `feature/parameter_quality`, I get the following errors when building in Windows 10:
grafik

**Comment by pablogs9** [comment]
Thanks, @norro, can you copy-paste the output?

@JanStaschulat that seems to be in the executor test also...

**Comment by pablogs9** [comment]
Those functions are not related to `rclc_parameter`, we should check the Windows compatibility of the rest of the package @JanStaschulat

**Comment by JanStaschulat** [comment]
Hmm, all these functions `rclc_executor_remove_*`have been added later. I'll check, I remember from my first port of rclc to Windows this has to do with visibility control. The macro `RCLC_PUBLIC` is missing infront of those functions. I'll update that.

**Comment by JanStaschulat** [comment]
@Arne, could you test it on Windows?

**Comment by norro** [comment]
Build of rclc_parameters in  126a7628deb3 fails on Windows 10 with

```
test_parameter.obj : error LNK2019: unresolved external symbol __imp_rclc_parameter_server_init_default referenced in function "private: virtual void __cdecl Test_rclc_node_init_default_Test::TestBody(void)" (?TestBody@Test_rclc_node_init_default_Test@@EEAAXXZ) [C:\Users\compass\ros2_ws\build\rclc_parameter\rclc_parameter_test.vcxproj]
test_parameter.obj : error LNK2019: unresolved external symbol __imp_rclc_executor_add_parameter_server referenced in function "private: virtual void __cdecl Test_rclc_node_init_default_Test::TestBody(void)" (?TestBody@Test_rclc_node_init_default_Test@@EEAAXXZ) [C:\Users\compass\ros2_ws\build\rclc_parameter\rclc_parameter_test.vcxproj]
test_parameter.obj : error LNK2019: unresolved external symbol __imp_rclc_add_parameter referenced in function "private: virtual void __cdecl Test_rclc_node_init_default_Test::TestBody(void)" (?TestBody@Test_rclc_node_init_default_Test@@EEAAXXZ) [C:\Users\compass\ros2_ws\build\rclc_parameter\rclc_parameter_test.vcxproj]
test_parameter.obj : error LNK2019: unresolved external symbol __imp_rclc_parameter_set_bool referenced in function "private: virtual void __cdecl Test_rclc_node_init_default_Test::TestBody(void)" (?TestBody@Test_rclc_node_init_default_Test@@EEAAXXZ) [C:\Users\compass\ros2_ws\build\rclc_parameter\rclc_parameter_test.vcxproj]
test_parameter.obj : error LNK2019: unresolved external symbol __imp_rclc_parameter_set_int referenced in function "private: virtual void __cdecl Test_rclc_node_init_default_Test::TestBody(void)" (?TestBody@Test_rclc_node_init_default_Test@@EEAAXXZ) [C:\Users\compass\ros2_ws\build\rclc_parameter\rclc_parameter_test.vcxproj]
test_parameter.obj : error LNK2019: unresolved external symbol __imp_rclc_parameter_set_double referenced in function "private: virtual void __cdecl Test_rclc_node_init_default_Test::TestBody(void)" (?TestBody@Test_rclc_node_init_default_Test@@EEAAXXZ) [C:\Users\compass\ros2_ws\build\rclc_parameter\rclc_parameter_test.vcxproj]
test_parameter.obj : error LNK2019: unresolved external symbol __imp_rclc_parameter_get_bool referenced in function "private: virtual void __cdecl Test_rclc_node_init_default_Test::TestBody(void)" (?TestBody@Test_rclc_node_init_default_Test@@EEAAXXZ) [C:\Users\compass\ros2_ws\build\rclc_parameter\rclc_parameter_test.vcxproj]
test_parameter.obj : error LNK2019: unresolved external symbol __imp_rclc_parameter_get_int referenced in function "private: virtual void __cdecl Test_rclc_node_init_default_Test::TestBody(void)" (?TestBody@Test_rclc_node_init_default_Test@@EEAAXXZ) [C:\Users\compass\ros2_ws\build\rclc_parameter\rclc_parameter_test.vcxproj]
test_parameter.obj : error LNK2019: unresolved external symbol __imp_rclc_parameter_get_double referenced in function "private: virtual void __cdecl Test_rclc_node_init_default_Test::TestBody(void)" (?TestBody@Test_rclc_node_init_default_Test@@EEAAXXZ) [C:\Users\compass\ros2_ws\build\rclc_parameter\rclc_parameter_test.vcxproj]
C:\Users\compass\ros2_ws\build\rclc_parameter\Release\rclc_parameter_test.exe : fatal error LNK1120: 9 unresolved externals [C:\Users\compass\ros2_ws\build\rclc_parameter\rclc_parameter_test.vcxproj]
Failed   <<< rclc_parameter [6.84s, exited with code 1]

Summary: 2 packages finished [18.8s]
  1 package failed: rclc_parameter
```

**Comment by pablogs9** [comment]
Those new symbols have the `RCLC_PARAMETER_PUBLIC` set. Any idea on how to solve it @JanStaschulat? Thanks!

**Comment by pablogs9** [comment]
Maybe e5bb567b1dd5a98af78df35153e3b0635fbb27d2? Can you test it again @norro? Sorry for the inconvenience...

**Comment by JanStaschulat** [comment]
not yet: 
```
  cl /c /IC:\dev\ros2_galactic\src\gtest_vendor\include /IC:\Users\compass\ros2_ws\src\rclc\rclc_parameter\include /IC:\Users\compass\ros2_ws\src\rclc\rclc_parameter\src /IC:\dev\ros2_galactic\include /IC:\Users\compass\ros2_ws\install\include /W3 /WX- /diagnostics:column /O2 /Ob2 /D WIN32 /D _WINDOWS /D NDEBUG /D RCLC_PARAMETER_MAX_STRING_LENGTH=50 /D _CRT_SECURE_NO_WARNINGS /D DEFAULT_RMW_IMPLEMENTATION=rmw_cyclonedds_cpp /D _SILENCE_ALL_CXX17_DEPRECATION_WARNINGS /D "CMAKE_INTDIR=\"Release\"" /D _MBCS /Gm- /EHsc /MD /GS /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /GR /std:c++17 /Fo"rclc_parameter_test.dir\Release\\" /Fd"rclc_parameter_test.dir\Release\vc142.pdb" /external:env:EXTERNAL_INCLUDE /external:W3 /Gd /TP /errorReport:queue C:\Users\compass\ros2_ws\src\rclc\rclc_parameter\test\rclc_parameter\test_parameter.cpp
  test_parameter.cpp
test_parameter.obj : error LNK2019: unresolved external symbol __imp_rclc_parameter_server_init_default referenced in function "private: virtual void __cdecl Test_rclc_node_init_default_Test::TestBody(void)" (?TestBody@Test_rclc_node_init_default_Test@@EEAAXXZ) [C:\Users\compass\ros2_ws\build\rclc_parameter\rclc_parameter_test.vcxproj]
test_parameter.obj : error LNK2019: unresolved external symbol __imp_rclc_executor_add_parameter_server referenced in function "private: virtual void __cdecl Test_rclc_node_init_default_Test::TestBody(void)" (?TestBody@Test_rclc_node_init_default_Test@@EEAAXXZ) [C:\Users\compass\ros2_ws\build\rclc_parameter\rclc_parameter_test.vcxproj]
test_parameter.obj : error LNK2019: unresolved external symbol __imp_rclc_add_parameter referenced in function "private: virtual void __cdecl Test_rclc_node_init_default_Test::TestBody(void)" (?TestBody@Test_rclc_node_init_default_Test@@EEAAXXZ) [C:\Users\compass\ros2_ws\build\rclc_parameter\rclc_parameter_test.vcxproj]
test_parameter.obj : error LNK2019: unresolved external symbol __imp_rclc_parameter_set_bool referenced in function "private: virtual void __cdecl Test_rclc_node_init_default_Test::TestBody(void)" (?TestBody@Test_rclc_node_init_default_Test@@EEAAXXZ) [C:\Users\compass\ros2_ws\build\rclc_parameter\rclc_parameter_test.vcxproj]
test_parameter.obj : error LNK2019: unresolved external symbol __imp_rclc_parameter_set_int referenced in function "private: virtual void __cdecl Test_rclc_node_init_default_Test::TestBody(void)" (?TestBody@Test_rclc_node_init_default_Test@@EEAAXXZ) [C:\Users\compass\ros2_ws\build\rclc_parameter\rclc_parameter_test.vcxproj]
test_parameter.obj : error LNK2019: unresolved external symbol __imp_rclc_parameter_set_double referenced in function "private: virtual void __cdecl Test_rclc_node_init_default_Test::TestBody(void)" (?TestBody@Test_rclc_node_init_default_Test@@EEAAXXZ) [C:\Users\compass\ros2_ws\build\rclc_parameter\rclc_parameter_test.vcxproj]
test_parameter.obj : error LNK2019: unresolved external symbol __imp_rclc_parameter_get_bool referenced in function "private: virtual void __cdecl Test_rclc_node_init_default_Test::TestBody(void)" (?TestBody@Test_rclc_node_init_default_Test@@EEAAXXZ) [C:\Users\compass\ros2_ws\build\rclc_parameter\rclc_parameter_test.vcxproj]
test_parameter.obj : error LNK2019: unresolved external symbol __imp_rclc_parameter_get_int referenced in function "private: virtual void __cdecl Test_rclc_node_init_default_Test::TestBody(void)" (?TestBody@Test_rclc_node_init_default_Test@@EEAAXXZ) [C:\Users\compass\ros2_ws\build\rclc_parameter\rclc_parameter_test.vcxproj]
test_parameter.obj : error LNK2019: unresolved external symbol __imp_rclc_parameter_get_double referenced in function "private: virtual void __cdecl Test_rclc_node_init_default_Test::TestBody(void)" (?TestBody@Test_rclc_node_init_default_Test@@EEAAXXZ) [C:\Users\compass\ros2_ws\build\rclc_parameter\rclc_parameter_test.vcxproj]
C:\Users\compass\ros2_ws\build\rclc_parameter\Release\rclc_parameter_test.exe : fatal error LNK1120: 9 unresolved externals [C:\Users\compass\ros2_ws\build\rclc_parameter\rclc_parameter_test.vcxproj]
Failed   <<< rclc_parameter [16.2s, exited with code 1]

Summary: 0 packages finished [16.5s]
  1 package failed: rclc_parameter

c:\Users\compass\ros2_ws>
```

**Comment by pablogs9** [comment]
No idea what is happening. Can you provide some tips @JanStaschulat?

**Comment by JanStaschulat** [comment]
we have been working on it. 
almost fixed

**Comment by pablogs9** [comment]
Thanks a lot

**Comment by JanStaschulat** [comment]
@pablogs9 can you sign your commit? I tried to signoff the last three commits, but Github expects you to sign your commit.
https://github.com/ros2/rclc/pull/144/checks?check_run_id=3321167322

**Comment by pablogs9** [comment]
I don't understand this:

image

I'm clicking Set DCO to pass

**Comment by JanStaschulat** [comment]
There is a new commit on `rcl` which lead to a failing CI job: 
https://github.com/ros2/rclc/runs/3321167476?check_suite_focus=true

Wrote a comment here: 
https://github.com/ros2/rcl/commit/4296dd1e94c444e4fca8b76a638eba80de27614d#diff-d463a4338ef6ff6cec2dada3378edb73e34fc476bebdd75ee5f25ed69fb4bd17

**Review by JanStaschulat — APPROVED** [review]

**Comment by JanStaschulat** [comment]
This version of `rclc_parameters` compiles and links on Windows. The units tests are all successfull. From my side, you can merge this pull request.

**Comment by pablogs9** [comment]
Nice, thanks a lot for the help. Merging.

**Comment by pablogs9** [comment]
@mergify backport galactic

**Comment by JanStaschulat** [comment]
Backport to Foxy we have to do manually, I guess.

**Comment by pablogs9** [comment]
We do not have parameters in Foxy

**Comment by JanStaschulat** [comment]
Created a selected backport for Foxy https://github.com/ros2/rclc/pull/182
